hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionScannerImpl.java

```diff
@@ -426,6 +426,8 @@ private boolean nextInternal(List<Cell> results, ScannerContext scannerContext)
     // Used to check time limit
     LimitScope limitScope = LimitScope.BETWEEN_CELLS;
 
+    checkpoint(State.START);
```

Comment:
It is just for RegionScannerImpl.filter  is not null we should checkpoint?

Reply:
good point. i think i can do that, let me look into it.

Reply:
I've moved the initial checkpoint call into the constructor, so we only have to do it once. The other checkpoint calls have been changed to a new `checkpointIfFiltering` method which returns early if `filter == null`.

The initial checkpoint call is still important because it enables `retainBlock()` functionality in StoreScanner. Someone could submit a scan with `addColumn(...)` which looks for 1 column in rows with many columns. In which case `retainBlock()` in StoreScanner would still be very useful. I added a comment to explain that.

Thanks for looking!

Reply:
@bbeaudreault , thank you very much for detailed reply. Overall LGTM, it is  is really a very insightful pr, just have a smoen suggestions , FYI.
Should we set  `HFileScannerImpl.lastCheckpointIndex` to  0 when initializing ? so we could simplify the 
    `if (shouldRetainBlock || lastCheckpointIndex < 0)`  in `HFileScannerImpl.handlePrevBlock` to 
`if (shouldRetainBlock)`, after all, when we start to scan,
`HFileScannerImpl.lastCheckpointIndex` is always >=0, no matter there is filter or not,  and we could also remove
`checkpoint(State.START)` in `RegionScannerImpl` ctor.

Reply:
@comnetwork I just pushed a commit which adds a `boolean checkpointingEnabled` to the creation of HFileScannerImpl.  This involves minor modifications to many levels of `getScanner(...)` calls.  I added the new param everywhere necessary, defaulting to `false` everywhere. This retains the old behavior for all usages. I then updated just StoreScanner to pass `true`, as this is the only place we want to enable this behavior right now.

This is a bunch more small changes, but is probably the safest route. If there were a bug where we accidentally passed false, we'd just be reverting to the original behavior and losing the optimization. You can see how it affects things in HFileReaderImpl.

Let me know what you think. I can revert that commit or we can keep it. As a result I was able to simplify some of the checkpointing (default to 0 instead of -1, no need to call checkpoint on new scanners).

Reply:
@bbeaudreault , thank you for the elaborate work and I agree your point that adding a new boolean `checkpointEnabled` make the logic more clear.

> The contract is that if you are using checkpointing, you also need to call retainBlock(). If you are not using checkpointing, you don't need to call retainBlock().

I agree with the first part of the sentence, but have doubts about the second. From you code, I think for scan by `RegionScannnerImpl`,  `retainBlock` is always needed to release block early,  only when there is filter(especially row level filter),we need checkpoint further to narrow the blocks which should be retained？So  we could only use `retainBlock`  for  user scan which specifying columns but not has filters.  If you agree with my point, I think the variable name `boolean checkpointingEnabled` is not very appropriate， I think maybe `earlyReleaseBlockEnabled` is more  indicative of the intentions？